pkgs/jnigen/doc/lifecycle.md

```diff
@@ -0,0 +1,150 @@
+[`Arena`](https://pub.dev/documentation/ffi/latest/ffi/Arena-class.html)s.
+First, create an `Arena` via
+[`using`](https://pub.dev/documentation/ffi/latest/ffi/using.html). Then
+register the object to be released at the end of the callback.
+
+```dart
+using((arena) {
+  final hello = 'Hello'.toJString()..releasedBy(arena);
+  final world = 'World'.toJString()..releasedBy(arena);
+  print(hello);
+  print(world);
+});
+// Both `hello` and `world` are now released.
+```
+
+### Lifecycle of interfaces implemented in Dart
+
+Java interfaces can be implemented in Dart using `.implement`. The Dart closure
+implementations must remain accessible as long as the created Java object
+exists. Ownership of these Dart objects is transferred to Java. When the Java GC
+reclaims the implemented object, a message is sent to Dart to remove the
+corresponding closures.
+
+> [!WARNING]
+>
+> One could create cycles between Dart and Java GC's when implementing
+> interfaces. For example consider the following:
+>
+> ```dart
+> final foo = Foo();
+> foo.bar = Bar.implement($Bar(
+>   f: () {
+>    return foo;
+>   }
+> ));
+> ```
+>
+> In Dart, closure `f` holds a reference to `foo`, preventing its garbage
+> collection. In Java, `foo` holds a reference to `bar`, preventing its release
+> and thus preventing the closure's removal. This creates a cycle.
+>
+> ```mermaid
+> graph TD;
+>    f-->|holds in Dart|foo;
+>    foo-->|holds in Java|bar;
+>    bar-->|holds in Java|f;
+> ```
+
+To prevent cycles, use
+[`WeakReference`](https://api.dart.dev/dart-core/WeakReference-class.html)s.
+
+```dart
+final weakFoo = WeakReference(foo);
+foo.bar = Bar.implement($Bar(
+  f: () {
+    final foo = weakFoo.target;
+    if (foo == null) {
+      throw StateError();
+    }
+    return foo;
+  }
+));
+```
+
+The weak reference breaks the cycle, allowing the GCs to collect the objects.
+
+```mermaid
+graph TD;
+   f-.->|holds weakly in Dart|foo;
+   foo-->|holds in Java|bar;
+   bar-->|holds in Java|f;
+```
+
+> [!WARNING]
+>
+> Be careful about the closures accidentally overcapturing. To prevent
+> overcapturing, implement your logic in a separate function or create a class
+> that implements `$Bar`.
+
+### Tips on keeping the number of global references low
+
+- Avoid storing `JObject`s in Dart collections like `List` or `Map`. Use Java
+  collections such as `JList` or `JMap` instead.
+
+  ```dart
+  // GOOD:
+  final jstrings = JList(JString.type);
+  using((arena) {
+    final hello = 'Hello'.toJString()..releasedBy(arena);
+    final world = 'World'.toJString()..releasedBy(arena);
+    jstrings.add(hello); // Add to Java collection
+    jstrings.add(world); // Add to Java collection
+  });
+  print(jstrings.length); // prints 2.
+  ```
+
+This approach avoids storing individual references. References are created only
+when accessing elements and released afterward.
+
+- Minimize the use of `static` or global `JObject`s to prevent them from being
```

Comment:
This is a bit ambiguous. Do you mean
"Minimize (the use of `static` or global `JObject`s to prevent them from being reclaimed)."
or
"(Minimize the use of `static` or global `JObject`s) to prevent them from being reclaimed."

I'm guessing it's the first one, so maybe rephrase it something like "It's tempting to prevent `JObject`s from being reclaimed by storing them in `static` or global variables. This pattern works well enough, but can use a lot of global references. So it's best to minimize this as much as possible".

Reply:
Yeah, the AI suggestion made it less clear. Fixed.